### ISSUE TITLE
cmake.selectActiveFolder command does not work if there are multiple workspace folders with the same name

### ISSUE THREAD

**Issue by DragosMiloiuNXP** [microsoft/vscode-cmake-tools#3538]
### Brief Issue Summary

Internally the cmake.selectActiveFolder command can receive the following parameter types:
- CMakeProject - unfortunately the CMake Tools API only allows me to get a Project, not a CMakeProject instance, so I cannot use the API to distinguish between multiple projects with the same name.
- the parameter can also be a list of strings, but these are the workspace folder names so it cannot differentiate between multiple folders with the same name.

**Comment by gcampbell-msft** [comment]
@DragosMiloiuNXP To help us understand what you'd like to see be improved, could you please detail what you are using this for, how you are currently interacting with that command, and what you'd like to be improved?

We're also happy to take open source contributions if you have a specific idea for a change. Thanks!

**Comment by gcampbell-msft** [comment]
@DragosMiloiuNXP I think I understand.

So, really, you're asking for two different types of possible fixes. 

1. If you still need to use `cmake.selectActiveFolder`, then either make it possible for API consumers to access to a CMakeProject to pass to this method, or define some other way to be able to distinguish between folders with the same name. 
2. OR, if you're interested in not using the `cmake.selectActiveFolder` at all, then make the task execution better at handling  

One last question: 

To clarify what you mean by this: 

> "Screenshot of the task execution without selectActiveFolder command. Task is correct, but the folder for which CMake is executed is incorrect."

You're saying that you were able to pick the right task to run, but it's running the task for the wrong folder, right? Does this end up meaning that the task is actually performing something you didn't want it to? For example, configuring a different project/folder then desired?

**Comment by DragosMiloiuNXP** [comment]
@gcampbell-msft,

> You're saying that you were able to pick the right task to run, but it's running the task for the wrong folder, right?

Yes. Yes it does the correct operation (build or clean) but for the wrong workspace folder (not the one associated with the task).

**Comment by TBRSTM** [comment]
I'm running into same issue. 
Currently no additional option in `tasks.json` to specify which sources we wish to target.

**Comment by alearq** [comment]
Hitting the same issue here, also when building code for NXP MCUs where the CMake files from the SDK projects are located in a folder called armgcc instead of the root folder.

Because of this my workspace contains two CMake project folders called "armgcc" and I cannot select the correct one using cmake.selectActiveFolder.

Even better would be if cmake tools could automatically select the correct cmake.sourceDirectory to build from based on which workspace the task was executed from.

**Comment by DragosMiloiuNXP** [comment]
Hi @alearq,
 
Regarding "Hitting the same issue here, also when building code for NXP MCUs where the CMake files from the SDK projects are located in a folder called armgcc instead of the root folder.", maybe you can try the MCUXpresso for VS Code extension. It is provided by NXP, it also relies on CMakeTools for CMake interaction, and it works with MCUXpresso ARM GCC based projects / SDK."

**Comment by alearq** [comment]
> Hi [@alearq](https://github.com/alearq),
> 
> Regarding "Hitting the same issue here, also when building code for NXP MCUs where the CMake files from the SDK projects are located in a folder called armgcc instead of the root folder.", maybe you can try the MCUXpresso for VS Code extension. It is provided by NXP, it also relies on CMakeTools for CMake interaction, and it works with MCUXpresso ARM GCC based projects / SDK."

I did use it initially to set up some projects but would prefer to use as little vendor-specific extensions as possible as my project involves more than one MCU on the same PCB from different manufacturers.

I did eventually find another workaround though. Turns out that putting multiple cmake.sourceDirectory entries in a code-workspace file like this:
```
"cmake.sourceDirectory": [
	"${workspaceFolder:projectA}",
	"${workspaceFolder:projectB}",
	"${workspaceFolder:projectC}/armgcc",
	"${workspaceFolder:projectD}/armgcc",
],
```
Creates the problem where cmake.autoSelectActiveFolder will only auto-select projectC and projectD if I open a file in the /armgcc folder and not if I open a file in that project but outside that directory, like /main.c. It also creates the problem that when executing the command "cmake.selectActiveFolder" from a VSCode task it won't be possible to distinguish between projectC and projectD because the only way to select both is by sending the input "armgcc" to it. Also if I skip the /armgcc part in the code-workspace file then the extension fails completely to find the cmake files.

But if I instead of putting multiple sourceDirectories in a code-workspace file put each in their own .vscode/settings.json file like this:
```
"cmake.sourceDirectory": "${workspaceFolder:projectC}/armgcc"
```

Then cmake.autoSelectActiveFolder will automatically select the correct project when opening any file from projectC and projectD **AND** the required input for the cmake.selectActiveFolder changes from "armgcc" to projectC and projectD respectively. So this turned out to solve my issue.
